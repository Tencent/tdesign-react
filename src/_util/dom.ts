import raf from 'raf';
import isString from 'lodash/isString';
import { easeInOutCubic, EasingFunction } from './easing';
import { ScrollContainer, ScrollContainerElement } from '../common';

// 用于判断是否可使用dom
export const canUseDocument = !!(typeof window !== 'undefined' && window.document && window.document.createElement);

const trim = (str: string): string => (str || '').replace(/^[\s\uFEFF]+|[\s\uFEFF]+$/g, '');

export const on = ((): any => {
  if (canUseDocument && document.addEventListener) {
    return (element: Node, event: string, handler: EventListenerOrEventListenerObject): any => {
      if (element && event && handler) {
        element.addEventListener(event, handler, false);
      }
    };
  }
  return (element: Node, event: string, handler: EventListenerOrEventListenerObject): any => {
    if (element && event && handler) {
      (element as any).attachEvent(`on${event}`, handler);
    }
  };
})();

export const off = ((): any => {
  if (canUseDocument && document.removeEventListener) {
    return (element: Node, event: string, handler: EventListenerOrEventListenerObject): any => {
      if (element && event) {
        element.removeEventListener(event, handler, false);
      }
    };
  }
  return (element: Node, event: string, handler: EventListenerOrEventListenerObject): any => {
    if (element && event) {
      (element as any).detachEvent(`on${event}`, handler);
    }
  };
})();

function hasClass(el: Element, cls: string) {
  if (!el || !cls) return false;
  if (cls.indexOf(' ') !== -1) throw new Error('className should not contain space.');
  if (el.classList) {
    return el.classList.contains(cls);
  }
  return ` ${el.className} `.indexOf(` ${cls} `) > -1;
}

export const addClass = function (el: Element, cls: string) {
  if (!el) return;
  let curClass = el.className;
  const classes = (cls || '').split(' ');

  for (let i = 0, j = classes.length; i < j; i++) {
    const clsName = classes[i];
    if (!clsName) continue;

    if (el.classList) {
      el.classList.add(clsName);
    } else if (!hasClass(el, clsName)) {
      curClass += ` ${clsName}`;
    }
  }
  if (!el.classList) {
    // eslint-disable-next-line
    el.className = curClass;
  }
};

export const removeClass = function (el: Element, cls: string) {
  if (!el || !cls) return;
  const classes = cls.split(' ');
  let curClass = ` ${el.className} `;

  for (let i = 0, j = classes.length; i < j; i++) {
    const clsName = classes[i];
    if (!clsName) continue;

    if (el.classList) {
      el.classList.remove(clsName);
    } else if (hasClass(el, clsName)) {
      curClass = curClass.replace(` ${clsName} `, ' ');
    }
  }
  if (!el.classList) {
    // eslint-disable-next-line
    el.className = trim(curClass);
  }
};

/**
 * 获取滚动容器
 * 因为document不存在scroll等属性, 因此排除document
 * window | HTMLElement
 * @param {ScrollContainerElement} [container='body']
 * @returns {ScrollContainer}
 */
export const getScrollContainer = (container: ScrollContainer = 'body'): ScrollContainerElement => {
  if (isString(container)) {
    return container ? (document.querySelector(container) as HTMLElement) : window;
  }
  if (typeof container === 'function') {
    return container();
  }
  return container;
};

/**
 * 返回是否window对象
 *
 * @export
 * @param {any} obj
 * @returns
 */
function isWindow(obj: any) {
  return obj && obj === obj.window;
}

type ScrollTarget = HTMLElement | Window | Document;

/**
 * 获取滚动距离
 *
 * @export
 * @param {ScrollTarget} target
 * @param {boolean} isLeft true为获取scrollLeft, false为获取scrollTop
 * @returns {number}
 */
export function getScroll(target: ScrollTarget, isLeft?: boolean): number {
  // node环境或者target为空
  if (typeof window === 'undefined' || !target) {
    return 0;
  }
  const method = isLeft ? 'scrollLeft' : 'scrollTop';
  let result = 0;
  if (isWindow(target)) {
    result = (target as Window)[isLeft ? 'pageXOffset' : 'pageYOffset'];
  } else if (target instanceof Document) {
    result = target.documentElement[method];
  } else if (target) {
    result = (target as HTMLElement)[method];
  }
  return result;
}
interface ScrollTopOptions {
  container?: ScrollTarget;
  duration?: number;
  easing?: EasingFunction;
}

export function scrollTo(target: number, opt: ScrollTopOptions) {
  const { container = window, duration = 450, easing = easeInOutCubic } = opt;
  const scrollTop = getScroll(container);
  const startTime = Date.now();
  return new Promise((res) => {
    const fnc = () => {
      const timestamp = Date.now();
      const time = timestamp - startTime;
      const nextScrollTop = easing(Math.min(time, duration), scrollTop, target, duration);
      if (isWindow(container)) {
        (container as Window).scrollTo(window.pageXOffset, nextScrollTop);
      } else if (container instanceof HTMLDocument || container.constructor.name === 'HTMLDocument') {
        (container as HTMLDocument).documentElement.scrollTop = nextScrollTop;
      } else {
        (container as HTMLElement).scrollTop = nextScrollTop;
      }
      if (time < duration) {
        raf(fnc);
      } else {
        // 由于上面步骤设置了scrollTop, 滚动事件可能未触发完毕
        // 此时应该在下一帧再执行res
        raf(res);
      }
    };
    raf(fnc);
  });
}

// 用于判断节点内容是否溢出
export const isNodeOverflow = (ele: Element | Element[]): boolean => {
  const { clientWidth = 0, scrollWidth = 0 } = ele as Element & { clientWidth: number; scrollWidth: number };

  if (scrollWidth > clientWidth) {
    return true;
  }
  return false;
};

export const getAttach = (node: any): HTMLElement => {
  const attachNode = typeof node === 'function' ? node() : node;
  if (!attachNode) {
    return document.body;
  }
  if (isString(attachNode)) {
    return document.querySelector(attachNode);
  }
  if (attachNode instanceof HTMLElement) {
    return attachNode;
  }
  return document.body;
};

// 获取 css vars
export const getCssVarsValue = (name: string, element?: HTMLElement) => {
  if (!canUseDocument) return;

  const el = element || document.documentElement;
  return getComputedStyle(el).getPropertyValue(name);
};
