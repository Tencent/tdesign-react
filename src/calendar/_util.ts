import dayjs from 'dayjs';
import { CalendarCell } from './type';

// 抽取配置常量
export const blockName = 'calendar'; // 类名前缀
export const controlSectionSize = 'medium'; // 操作栏控件尺寸
export const minYear = 1970; // 最早选择年份

/**
 * 获取一个日期是周几（1~7）
 */
export const getDay = (dt: Date): number => {
  let day = dayjs(dt).day();
  if (day === 0) {
    day = 7;
  }
  return day;
};

/**
 * 创建日历单元格数据
 * @param year 日历年份
 * @param month 日历月份
 * @param firstDayOfWeek 周起始日（1~7）
 * @param currentValue 当前日期
 * @param format 日期格式
 */
export const createDateList = (
  year: number,
  month: number,
  firstDayOfWeek: number,
  currentValue: dayjs.Dayjs,
  format: string,
): CalendarCell[][] => {
  const createCellData = (belongTo: number, isCurrent: boolean, date: Date, weekOrder: number): CalendarCell => {
    const day = getDay(date);
    return {
      mode: 'month',
      belongTo,
      isCurrent,
      day,
      weekOrder,
      date,
      formattedDate: dayjs(date).format(format),
      filterDate: null,
      formattedFilterDate: null,
      isShowWeekend: true,
    };
  };

  const monthFirstDay = dayjs(`${year}-${month}`);
  const rowList = [] as CalendarCell[][];
  let list = [] as CalendarCell[];
  let weekCount = 1;

  // 添加上个月中会在本月显示的最后几天日期
  const lastMonthDaysCount = (getDay(monthFirstDay.toDate()) - firstDayOfWeek + 7) % 7;
  for (let i = 0; i < lastMonthDaysCount; i++) {
    const dayObj = monthFirstDay.subtract(i + 1, 'day');
    list.unshift(createCellData(-1, false, dayObj.toDate(), weekCount));
  }

  // 添加本月日期
  const monthDaysCount = monthFirstDay.endOf('month').daysInMonth();
  for (let i = 0; i < monthDaysCount; i++) {
    const dayObj = monthFirstDay.add(i, 'day');
    list.push(createCellData(0, currentValue.isSame(dayObj), dayObj.toDate(), weekCount));
    if (list.length === 7) {
      rowList.push(list);
      list = [];
      weekCount += 1;
    }
  }

  // 添加下月日期
  if (list.length) {
    const monthLastDay = dayjs(`${year}-${month}`).endOf('month');
    const nextMonthDaysCount = 7 - list.length;
    for (let i = 0; i < nextMonthDaysCount; i++) {
      const dayObj = monthLastDay.add(i + 1, 'day');
      list.push(createCellData(1, false, dayObj.toDate(), weekCount));
    }
    rowList.push(list);
  }

  return rowList;
};

/**
 * 创建月历单元格数据
 * @param year 月历年份
 * @param currentValue 当前日期
 * @param format 输出格式
 * @param rowNum 一行日历的个数
 */
export const createMonthList = (
  year: number,
  currentValue: dayjs.Dayjs,
  format: string,
  rowNum = 4,
): CalendarCell[][] => {
  const monthsArr: CalendarCell[][] = [];
  const isCurYear = currentValue.year() === year;
  let tmpArr: CalendarCell[] = [];
  for (let i = 1; i <= 12; i++) {
    const date = new Date(year, i - 1);
    const curDateMon = parseInt(currentValue.format('M'), 10);
    const isCurrent = isCurYear && curDateMon === i;

    tmpArr.push({
      mode: 'year',
      isCurrent,
      date,
      formattedDate: dayjs(date).format(format),
      filterDate: null,
      formattedFilterDate: null,
      isShowWeekend: true,
    });
    if (tmpArr.length === rowNum) {
      monthsArr.push(tmpArr);
      tmpArr = [];
    }
  }

  return monthsArr;
};
