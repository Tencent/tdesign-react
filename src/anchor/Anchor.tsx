import React, { useState, useRef, useEffect, useCallback, useImperativeHandle } from 'react';
import classNames from 'classnames';
import isEmpty from 'lodash/isEmpty';
import isFunction from 'lodash/isFunction';
import { StyledProps } from '../common';
import { TdAnchorProps } from './type';
import useConfig from '../hooks/useConfig';
import forwardRefWithStatics from '../_util/forwardRefWithStatics';
import { canUseDocument, getScrollContainer } from '../_util/dom';
import Affix from '../affix';
import { AnchorContext, Item } from './AnchorContext';
import { getOffsetTop, getScroll, scrollTo, AnchorContainer } from './_util/dom';
import AnchorItem from './AnchorItem';
import AnchorTarget from './AnchorTarget';
import { anchorDefaultProps } from './defaultProps';

export interface AnchorProps extends TdAnchorProps, StyledProps {
  children?: React.ReactNode;
}

interface IntervalRef {
  // 收集 anchor-item
  items: string[];
  // 容器
  scrollContainer: AnchorContainer;
  // 收集各项 item 的信息与节点
  handleScrollLock: boolean;
}

const ANCHOR_SHARP_REGEXP = /#(\S+)$/;

const Anchor = forwardRefWithStatics(
  (props: AnchorProps, ref: React.Ref<HTMLDivElement>) => {
    const {
      affixProps,
      bounds,
      targetOffset,
      container,
      size,
      children,
      cursor,
      onClick,
      onChange,
      className,
      ...rest
    } = props;

    const { classPrefix } = useConfig();

    const [activeItem, setActiveItem] = useState<string>('');
    const [cursorStyle, setCursorStyle] = useState<{ top: string; height?: string; opacity: number }>({
      top: '0px',
      height: '0px',
      opacity: 0,
    });

    const anchorEl = useRef(null);
    const intervalRef = useRef<IntervalRef>({
      items: [],
      scrollContainer: canUseDocument ? window : null,
      handleScrollLock: false,
    });

    useImperativeHandle(ref, () => anchorEl.current);

    /**
     * 注册锚点
     * @param href 链接
     */
    const registerItem = (href: string): void => {
      const { items } = intervalRef.current;
      if (ANCHOR_SHARP_REGEXP.test(href) && items.indexOf(href) < 0) items.push(href);
    };

    const unregisterItem = (href: string): void => {
      const { items } = intervalRef.current;
      intervalRef.current.items = items.filter((item) => href !== item);
    };

    const getAnchorTarget = (href: string): HTMLElement => document.querySelector(href);

    const handleScrollTo = (link: string) => {
      const anchor = getAnchorTarget(link);
      if (!anchor) return;
      onChange?.(link, activeItem);
      setActiveItem(link);
      intervalRef.current.handleScrollLock = true;
      const { scrollContainer } = intervalRef.current;
      const scrollTop = getScroll(scrollContainer);
      const offsetTop = getOffsetTop(anchor, scrollContainer);
      const top = scrollTop + offsetTop - targetOffset;
      scrollTo(top, {
        container: scrollContainer,
      }).then(() => {
        intervalRef.current.handleScrollLock = false;
      });
    };

    const handleClick = (item: Item, e: React.MouseEvent<HTMLDivElement>) => {
      onClick?.({ e, ...item });
      handleScrollTo(item.href);
    };

    useEffect(() => {
      // update point style
      const pointEl = anchorEl.current.querySelector?.(`.${classPrefix}-is-active>a`) as HTMLAnchorElement;
      if (!pointEl) {
        setCursorStyle(null);
      } else {
        const { offsetTop: top, offsetHeight: height } = pointEl;
        setCursorStyle({ top: `${top}px`, height: `${height}px`, opacity: 1 });
      }
    }, [activeItem, classPrefix]);

    const handleScroll = useCallback(() => {
      const { scrollContainer, handleScrollLock } = intervalRef.current;
      if (handleScrollLock) return;
      const { items } = intervalRef.current;
      const filters: { top: number; href: string }[] = [];
      let active = '';
      // 找出所有当前 top 小于预设值
      items.forEach((href) => {
        const anchor = getAnchorTarget(href);
        if (!anchor) return;
        const top = getOffsetTop(anchor, scrollContainer);
        if (top <= bounds + targetOffset) {
          filters.push({
            href,
            top,
          });
        }
      });

      // 找出小于预设值集合中top最大的
      if (filters.length) {
        const latest = filters.reduce((prev, cur) => (prev.top > cur.top ? prev : cur));
        active = latest.href;
      }
      if (active !== activeItem) {
        onChange?.(active, activeItem);
        setActiveItem(active);
      }
    }, [activeItem, bounds, onChange, targetOffset]);

    useEffect(() => {
      intervalRef.current.scrollContainer = getScrollContainer(container);
      const { scrollContainer } = intervalRef.current;

      handleScroll();
      scrollContainer?.addEventListener('scroll', handleScroll);
      return () => {
        scrollContainer?.removeEventListener('scroll', handleScroll);
      };
    }, [container, handleScroll]);

    const anchorClass = classNames(
      `${classPrefix}-anchor`,
      {
        [`${classPrefix}-size-s`]: size === 'small',
        [`${classPrefix}-size-m`]: size === 'medium',
        [`${classPrefix}-size-l`]: size === 'large',
      },
      className,
    );

    const CursorCmp = () => {
      if (isFunction(cursor)) return cursor();
      if (isEmpty(cursor)) return <div className={`${classPrefix}-anchor__line-cursor`}></div>;
      return cursor;
    };

    const Cmp = (
      <AnchorContext.Provider
        value={{
          onClick: handleClick,
          activeItem,
          registerItem,
          unregisterItem,
        }}
      >
        <div {...rest} className={anchorClass} ref={anchorEl}>
          <div className={`${classPrefix}-anchor__line`}>
            <div className={`${classPrefix}-anchor__line-cursor-wrapper`} style={cursorStyle}>
              {CursorCmp()}
            </div>
          </div>
          {children}
        </div>
      </AnchorContext.Provider>
    );

    return isEmpty(affixProps) ? Cmp : <Affix {...affixProps}>{Cmp}</Affix>;
  },
  {
    AnchorItem,
    AnchorTarget,
  },
);

Anchor.displayName = 'Anchor';
Anchor.defaultProps = anchorDefaultProps;

export default Anchor;
