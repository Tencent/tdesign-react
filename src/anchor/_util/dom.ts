import raf from 'raf';
import { easeInOutCubic, EasingFunction } from '../../_util/easing';

export type AnchorContainer = HTMLElement | Window;

export function getOffsetTop(target: HTMLElement, container: AnchorContainer) {
  const { top } = target.getBoundingClientRect();
  if (container === window) {
    // 减去 document 的边框
    return top - document.documentElement.clientTop;
  }
  return top - (container as HTMLElement).getBoundingClientRect().top;
}

/**
 * 返回是否 window 对象
 *
 * @export
 * @param {any} obj
 * @returns
 */
function isWindow(obj: any) {
  return obj && obj === obj.window;
}

type ScrollTarget = HTMLElement | Window | Document;

/**
 * 获取滚动距离
 *
 * @export
 * @param {ScrollTarget} target
 * @param {boolean} isLeft true为获取scrollLeft, false为获取scrollTop
 * @returns {number}
 */
export function getScroll(target: ScrollTarget, isLeft?: boolean): number {
  // node 环境或者 target 为空
  if (typeof window === 'undefined' || !target) {
    return 0;
  }
  const method = isLeft ? 'scrollLeft' : 'scrollTop';
  let result = 0;
  if (isWindow(target)) {
    result = (target as Window)[isLeft ? 'pageXOffset' : 'pageYOffset'];
  } else if (target instanceof Document) {
    result = target.documentElement[method];
  } else if (target) {
    result = (target as HTMLElement)[method];
  }
  return result;
}

interface ScrollTopOptions {
  container?: ScrollTarget;
  duration?: number;
  easing?: EasingFunction;
}
export function scrollTo(target: number, opt: ScrollTopOptions) {
  const { container = window, duration = 450, easing = easeInOutCubic } = opt;
  const scrollTop = getScroll(container);
  const startTime = Date.now();
  return new Promise((res) => {
    const fnc = () => {
      const timestamp = Date.now();
      const time = timestamp - startTime;
      const nextScrollTop = easing(Math.min(time, duration), scrollTop, target, duration);
      if (isWindow(container)) {
        (container as Window).scrollTo(window.pageXOffset, nextScrollTop);
      } else if (container instanceof HTMLDocument || container.constructor.name === 'HTMLDocument') {
        (container as HTMLDocument).documentElement.scrollTop = nextScrollTop;
      } else {
        (container as HTMLElement).scrollTop = nextScrollTop;
      }
      if (time < duration) {
        raf(fnc);
      } else {
        // 由于上面步骤设置了 scrollTop, 滚动事件可能未触发完毕
        // 此时应该在下一帧再执行 res
        raf(res);
      }
    };
    raf(fnc);
  });
}
