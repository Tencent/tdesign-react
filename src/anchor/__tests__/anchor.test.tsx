import React from 'react';
import { testExamples, render, fireEvent } from '@test/utils';
import Anchor from '../Anchor';
import { getScroll, scrollTo } from '../_util/dom';

const { AnchorTarget, AnchorItem } = Anchor;

// 测试组件代码 Example 快照
testExamples(__dirname);

jest.resetModules();

describe('Anchor', () => {
  it('util/clipboard', () => {
    jest.mock('../_util/clipboard', () => (args) => args);
    const wrapper = render(
      <AnchorTarget id="/components/anchor/#default" tag="h1">
        基础锚点
      </AnchorTarget>,
    );
    const a = wrapper.container.querySelector('.t-icon-file-copy');
    fireEvent.click(a);

    expect(true).toEqual(true);
  });

  it('util/dom', () => {
    const distance = 0;
    const childTestID = 'childTestID';
    jest.mock('../_util/dom', () => ({
      getScroll: () => 0,
      scrollTo: (args) => args,
    }));

    const wrapper = render(
      <Anchor targetOffset={150}>
        <AnchorItem href="#基础锚点" title="基础锚点" data-testid={childTestID} />
      </Anchor>,
    );

    // target is element
    const domDistance = getScroll(wrapper.getByTestId(childTestID), true);
    expect(domDistance).toEqual(distance);

    // target is null
    const nullDistance = getScroll(null, true);
    expect(nullDistance).toEqual(distance);

    // target is document
    const documentDistance = getScroll(document, true);
    expect(documentDistance).toEqual(distance);

    // target is window
    const windowDistance = getScroll(window, true);
    expect(windowDistance).toEqual(distance);

    const scrollToPromise = scrollTo(10, {});
    expect(scrollToPromise).toBeInstanceOf(Promise);
  });
});
