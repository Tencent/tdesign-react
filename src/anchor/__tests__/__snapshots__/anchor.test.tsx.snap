// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base.jsx 1`] = `
<DocumentFragment>
  <div
    style="display: flex;"
  >
    <div
      style="flex: 1;"
    />
    <div>
      <div>
        <div
          class="t-anchor t-size-m"
        >
          <div
            class="t-anchor__line"
          >
            <div
              class="t-anchor__line-cursor-wrapper"
              style=""
            >
              <div
                class="t-anchor__line-cursor"
              />
            </div>
          </div>
          <div
            class="t-anchor__item"
          >
            <a
              class="t-anchor__item-link"
              href="#基础锚点"
              target="_self"
              title="基础锚点"
            >
              基础锚点
            </a>
          </div>
          <div
            class="t-anchor__item"
          >
            <a
              class="t-anchor__item-link"
              href="#多级锚点"
              target="_self"
              title="多级锚点"
            >
              多级锚点
            </a>
          </div>
          <div
            class="t-anchor__item"
          >
            <a
              class="t-anchor__item-link"
              href="#指定容器锚点"
              target="_self"
              title="指定容器锚点"
            >
              指定容器锚点
            </a>
          </div>
          <div
            class="t-anchor__item"
          >
            <a
              class="t-anchor__item-link"
              href="#特定交互锚点"
              target="_self"
              title="特定交互锚点"
            >
              特定交互锚点
            </a>
          </div>
          <div
            class="t-anchor__item"
          >
            <a
              class="t-anchor__item-link"
              href="#尺寸"
              target="_self"
              title="尺寸"
            >
              尺寸
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`container.jsx 1`] = `
<DocumentFragment>
  <div
    class="anchor-demo-attach"
    id="attach"
    style="display: flex;"
  >
    <div
      class="t-anchor t-size-m"
    >
      <div
        class="t-anchor__line"
      >
        <div
          class="t-anchor__line-cursor-wrapper"
          style="top: 0px; height: 0px; opacity: 1;"
        >
          <div
            class="t-anchor__line-cursor"
          />
        </div>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#content-1"
          target="_self"
          title="content-1"
        >
          content-1
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#content-2"
          target="_self"
          title="content-2"
        >
          content-2
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#content-3"
          target="_self"
          title="content-3"
        >
          content-3
        </a>
      </div>
      <div
        class="t-anchor__item t-is-active"
      >
        <a
          class="t-anchor__item-link"
          href="#content-4"
          target="_self"
          title="content-4"
        >
          content-4
        </a>
      </div>
    </div>
    <div
      id="anchor-container"
      style="width: 100%; height: 200px; overflow: auto; text-align: center; font-size: 22px;"
    >
      <div
        id="content-1"
        style="background: rgb(223, 239, 255); line-height: 100px;"
      >
        content-1
      </div>
      <div
        id="content-2"
        style="background: rgb(191, 219, 247); line-height: 100px;"
      >
        content-2
      </div>
      <div
        id="content-3"
        style="background: rgb(155, 197, 242); line-height: 100px;"
      >
        content-3
      </div>
      <div
        id="content-4"
        style="background: rgb(123, 175, 237); line-height: 100px;"
      >
        content-4
      </div>
      <div
        id="content-5"
        style="background: rgb(92, 153, 235); line-height: 100px;"
      >
        content-5
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`cursor.jsx 1`] = `
<DocumentFragment>
  <div
    style="display: flex;"
  >
    <div
      class="t-anchor t-size-m"
    >
      <div
        class="t-anchor__line"
      >
        <div
          class="t-anchor__line-cursor-wrapper"
          style=""
        >
          <div
            style="width: 10px; height: 10px; background-color: rgb(0, 82, 217); position: absolute; border-radius: 50%; left: 50%; margin-left: -5px; top: 50%; margin-top: -5px;"
          />
        </div>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#基础锚点"
          target="_self"
          title="基础锚点"
        >
          基础锚点
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#多级锚点"
          target="_self"
          title="多级锚点"
        >
          多级锚点
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#指定容器锚点"
          target="_self"
          title="指定容器锚点"
        >
          指定容器锚点
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#特定交互锚点"
          target="_self"
          title="特定交互锚点"
        >
          特定交互锚点
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#尺寸"
          target="_self"
          title="尺寸"
        >
          尺寸
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`large.jsx 1`] = `
<DocumentFragment>
  <div
    id="size-large"
  >
    <div
      class="t-anchor t-size-l"
    >
      <div
        class="t-anchor__line"
      >
        <div
          class="t-anchor__line-cursor-wrapper"
          style="top: 0px; height: 0px; opacity: 1;"
        >
          <div
            class="t-anchor__line-cursor"
          />
        </div>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#default"
          target="_self"
          title="基础锚点"
        >
          基础锚点
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#multiple"
          target="_self"
          title="多级锚点"
        >
          多级锚点
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#size"
          target="_self"
          title="尺寸大小"
        >
          尺寸大小
        </a>
        <div
          class="t-anchor__item"
        >
          <a
            class="t-anchor__item-link"
            href="#size"
            target="_self"
            title="size-medium"
          >
            size-medium
          </a>
        </div>
        <div
          class="t-anchor__item t-is-active"
        >
          <a
            class="t-anchor__item-link"
            href="#size-large"
            target="_self"
            title="size-large"
          >
            size-large
          </a>
        </div>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#attach"
          target="_self"
          title="指定容器"
        >
          指定容器
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`multiple.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-anchor t-size-m"
  >
    <div
      class="t-anchor__line"
    >
      <div
        class="t-anchor__line-cursor-wrapper"
        style=""
      >
        <div
          class="t-anchor__line-cursor"
        />
      </div>
    </div>
    <div
      class="t-anchor__item"
    >
      <a
        class="t-anchor__item-link"
        href="#基础锚点"
        target="_self"
        title="基础锚点"
      >
        基础锚点
      </a>
    </div>
    <div
      class="t-anchor__item"
    >
      <a
        class="t-anchor__item-link"
        href="#多级锚点"
        target="_self"
        title="多级锚点"
      >
        多级锚点
      </a>
    </div>
    <div
      class="t-anchor__item"
    >
      <a
        class="t-anchor__item-link"
        href="#指定容器锚点"
        target="_self"
        title="指定容器锚点"
      >
        指定容器锚点
      </a>
    </div>
    <div
      class="t-anchor__item"
    >
      <a
        class="t-anchor__item-link"
        href="#特定交互锚点"
        target="_self"
        title="特定交互锚点"
      >
        特定交互锚点
      </a>
    </div>
    <div
      class="t-anchor__item"
    >
      <a
        class="t-anchor__item-link"
        href="#尺寸"
        target="_self"
        title="尺寸"
      >
        尺寸
      </a>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#size"
          target="_self"
          title="size-medium"
        >
          size-medium
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#size-large"
          target="_self"
          title="size-large"
        >
          size-large
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`small.jsx 1`] = `
<DocumentFragment>
  <div
    id="size-small"
  >
    <div
      class="t-anchor t-size-s"
    >
      <div
        class="t-anchor__line"
      >
        <div
          class="t-anchor__line-cursor-wrapper"
          style=""
        >
          <div
            class="t-anchor__line-cursor"
          />
        </div>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#default"
          target="_self"
          title="基础锚点"
        >
          基础锚点
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#multiple"
          target="_self"
          title="多级锚点"
        >
          多级锚点
        </a>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#size"
          target="_self"
          title="尺寸大小"
        >
          尺寸大小
        </a>
        <div
          class="t-anchor__item"
        >
          <a
            class="t-anchor__item-link"
            href="#size"
            target="_self"
            title="size-medium"
          >
            size-medium
          </a>
        </div>
        <div
          class="t-anchor__item"
        >
          <a
            class="t-anchor__item-link"
            href="#size-large"
            target="_self"
            title="size-large"
          >
            size-large
          </a>
        </div>
      </div>
      <div
        class="t-anchor__item"
      >
        <a
          class="t-anchor__item-link"
          href="#attach"
          target="_self"
          title="指定容器"
        >
          指定容器
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`target.jsx 1`] = `
<DocumentFragment>
  <div
    class="anchor-demo"
    id="/components/anchor/#anchor-target"
  >
    <h1
      class="t-anchor__target"
      id="/components/anchor/#default"
    >
      基础锚点
      <div
        class="t-popup__reference"
      >
        <svg
          class="t-icon t-icon-file-copy t-anchor__copy"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M4 1.92C4 1.34 4.52 1 5 1h4.37a1 1 0 01.71.3L13.71 5a1 1 0 01.29.7v6.38c0 .58-.52.92-1 .92H5c-.48 0-1-.34-1-.92V1.92zM5 2v10h8V6.01H9V2H5zm5 .65V5h2.32L10 2.65z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M2 5v9.01a1 1 0 001 1h8v-1H3V5H2z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
    </h1>
    <h1
      class="t-anchor__target"
      id="/components/anchor/#multiple"
    >
      多级锚点
      <div
        class="t-popup__reference"
      >
        <svg
          class="t-icon t-icon-file-copy t-anchor__copy"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M4 1.92C4 1.34 4.52 1 5 1h4.37a1 1 0 01.71.3L13.71 5a1 1 0 01.29.7v6.38c0 .58-.52.92-1 .92H5c-.48 0-1-.34-1-.92V1.92zM5 2v10h8V6.01H9V2H5zm5 .65V5h2.32L10 2.65z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M2 5v9.01a1 1 0 001 1h8v-1H3V5H2z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
    </h1>
    <h1
      class="t-anchor__target"
      id="/components/anchor/#size"
    >
      尺寸大小
      <div
        class="t-popup__reference"
      >
        <svg
          class="t-icon t-icon-file-copy t-anchor__copy"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M4 1.92C4 1.34 4.52 1 5 1h4.37a1 1 0 01.71.3L13.71 5a1 1 0 01.29.7v6.38c0 .58-.52.92-1 .92H5c-.48 0-1-.34-1-.92V1.92zM5 2v10h8V6.01H9V2H5zm5 .65V5h2.32L10 2.65z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M2 5v9.01a1 1 0 001 1h8v-1H3V5H2z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
    </h1>
    <h1
      class="t-anchor__target"
      id="/components/anchor/#attach"
    >
      指定容器
      <div
        class="t-popup__reference"
      >
        <svg
          class="t-icon t-icon-file-copy t-anchor__copy"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M4 1.92C4 1.34 4.52 1 5 1h4.37a1 1 0 01.71.3L13.71 5a1 1 0 01.29.7v6.38c0 .58-.52.92-1 .92H5c-.48 0-1-.34-1-.92V1.92zM5 2v10h8V6.01H9V2H5zm5 .65V5h2.32L10 2.65z"
            fill="currentColor"
            fill-opacity="0.9"
          />
          <path
            d="M2 5v9.01a1 1 0 001 1h8v-1H3V5H2z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
    </h1>
  </div>
</DocumentFragment>
`;
